File: 3rd Semester/Advanced Programming Methods(Done)/Laboratory/src (BUX)/Model/Statements/FileOperations/closeRFileStmt.java
package Model.Statements.FileOperations;

import Model.*;
import Model.Expressions.Exp;
import Exceptions.MyControllerException;
import Exceptions.MyStmtException;
import Model.Statements.IStmt;
import Model.Utils.IFileTable;
import Model.Utils.MyIDictionary;
import Model.Utils.MyIHeap;
import Model.Utils.Tuple;

import java.io.BufferedReader;
import java.io.IOException;

public class closeRFileStmt implements IStmt
{
    private Exp exp;

    public closeRFileStmt(Exp exp)
    {
        this.exp = exp;
    }
   public PrgState execute(PrgState state) throws MyControllerException {
        MyIDictionary<String, Integer> dict = state.getSymTable();
       MyIHeap<Integer> heap=state.getHeap();
        IFileTable<Tuple> tpl = state.getFT();

        try{
            int val = exp.eval(dict,heap);
            Tuple tup = tpl.get(val);
            if(tup != null)
            {
                BufferedReader bf = tup.getBf();
                bf.close();
                tpl.remove(val);
            }
            else
                throw new MyControllerException("The file does not exist!");
        }catch (MyStmtException e)
        {
            throw new MyControllerException(e);
        }catch (IOException e)
        {
            e.printStackTrace();
        }
        return null;
    }

    public String toString() {
        return "Closing file";
    }
}

File: 3rd Semester/Advanced Programming Methods(Done)/Laboratory/C#LabAssignment/sharplab/Model/Statements/closeRFileStmt.cs
using sharplab.Model.Expressions;
using sharplab.Model.Utils;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace sharplab.Model.Statements
{
    class CloseRFileStmt : IStmt
    {
        private Exp exp;

        public CloseRFileStmt(Exp exp)
        {
            this.exp = exp;
        }
        public PrgState Execute(PrgState state)
        {
            IMyDictionary<string, int> dict = state.GetSymTable();
            IFileTable<Utils.Tuple> tpl = state.GetFT();

            try
            {
                int val = exp.Eval(dict);
                Utils.Tuple tup = tpl.Get(val);
                if(tup != null)
                {
                    StreamReader s = tup.GetStream();
                    s.Close();
                    tpl.Remove(val);
                }
                else
                    throw new Exception("The file does not exist!");
            }catch 
            {
                throw new Exception("Invalid expression!");
            }
            return state;
        }

        public override string ToString()
        {
            return "Closing file";
        }
    }
}
